feat(search): adds a new core search service

hypeJunction opened the pull request:
Adds a new search service that behaves similar to elgg_get_entities()

BREAKING CHANGE
Most of search plugin views have been removed or updated
Tag search has been combined with entity search.
Some of the search behavior may have changed, search hooks have been deprecated.

Revives #10688

- [x] Add tests
- [x] ~~~Add group profile fields to search?~~~

jeabakker commented on views/default/page/elements/by_line.php:
```diff
@@ -0,0 +1,65 @@
+<?php
```
why this file? we already have object/elements/imprint/byline

hypeJunction replied:
Hm, merge conflicts

jeabakker commented:
I didn't see a way for a different search engine (Elasticsearch or Solar) to be able to provide the search results. Did I miss something?

If not, this should be able, because MySQL search s**cks

hypeJunction commented:
> because MySQL search s**cks

Yes, let's drop it ;)

So, the big question is whether we should just add indexing support at an ege* level. Working with Solr, I always want to just use it for ege with metadata, maybe we should just build that into `Entities::get` to allow plugins to serve entities without querying the database

jeabakker commented:
> So, the big question is whether we should just add indexing support at an ege* level.

No, that is about storage and yes that could be changed, but not as part of this change.


> maybe we should just build that into Entities::get to allow plugins to serve entities without querying the database

same: no (imo)

If the search service has a hook before elgg_get_entities than everything can be fixed

jdalsem commented:
> Add group profile fields to search?

i would not do this as part of this PR... can discuss this later

hypeJunction commented:
Ok, let me try and wrap this up. Afterwards there is only one more thing I would like to get into 3.0.

jdalsem commented on mod/search/start.php:
```diff
@@ -495,6 +686,49 @@ function search_output_tag(\Elgg\Hook $hook) {
+ * @param \Elgg\Hook $hook Hook
+ * @return \ElggEntity
+ */
+function search_format_comment_entity(\Elgg\Hook $hook) {
```
can't we move this to a view?

hypeJunction replied:
I want to stop doing crazy shit in views

jdalsem replied:
instead do crazy shit in functions :) not going to block on this

hypeJunction replied:
A function can be tested, view testability is a hassle

hypeJunction replied:
In a perfect world, views are templates, they should not be altering entities, setting/resetting global state, especially without cleaning up after themselves

jdalsem replied:
agree.... only wondering why this has to happen in the search plugin and not somewhere else

hypeJunction replied:
Because search plugin is the one using the volatile data

jdalsem replied:
ah, maybe the word `format` is a bit confusing as it only prepares volatile data on the entity...

jdalsem commented on mod/search/views/default/forms/search.php:
```diff
@@ -0,0 +1,54 @@
+ */
+$value = elgg_extract('value', $vars, get_input('q', get_input('tag')));
+
+$search_attrs = elgg_format_attributes(array(
```
[] ?

jdalsem commented:
Great PR... only thing i really need is the ability to influence the getter of the search... just add a hook and it would give plugin devs control over the results

hypeJunction commented on engine/classes/Elgg/Search/SearchService.php:
```diff
@@ -0,0 +1,315 @@
+
+		$options = $this->hooks->trigger('search:options', $search_type, $options, $options);
+
+		return elgg_get_entities($options);
```
Will do

jdalsem commented on mod/search/start.php:
```diff
@@ -495,6 +686,49 @@ function search_output_tag(\Elgg\Hook $hook) {
+	$container = $entity->getContainerEntity();
+
+	if ($container->getType() == 'object') {
+		$title = $container->title;
```
use `getDisplayName()`

jeabakker commented on docs/appendix/upgrade-notes/2.x-to-3.0.rst:
```diff
@@ -409,11 +429,11 @@ The following class registration related functions have been removed:
 Search changes
 --------------
 
+We have added a search service into core, consequently the ``search`` plugin now only provides a user interface for displaying forms and listing search results. May of the views in the search plugin have been affected by this change.
```
> May of the views in the search plugin..

Many

jeabakker commented on mod/search/classes/Elgg/Search/FormatComentEntityHook.php:
```diff
@@ -0,0 +1,58 @@
+			$title = elgg_echo('item:' . $container->getType() . ':' . $container->getSubtype());
+		}
+
+		if (!$title) {
```
can you get here, as an `elgg_echo()` of a non-existing language key results in the key being echoed

hypeJunction replied:
could you clarify?

jeabakker replied:
the `if` above this one (line 42) is almost the same, but it checks a different language key. So i think the first `if` (line 42) will make sure you can never get into this one.

jeabakker replied:
maybe if the language key exists but it's an empty string...

jdalsem commented on mod/search/classes/Elgg/Search/FormatComentEntityHook.php:
```diff
@@ -0,0 +1,58 @@
+
+		$title = $container->getDisplayName();
+
+		if (!$title) {
```
you probably meant to do `&& elgg_language_key_exists`

jdalsem commented on engine/classes/Elgg/Search/SearchService.php:
```diff
@@ -75,66 +75,110 @@ public function __construct(Config $config, PluginHooksService $hooks, Database
 
 		$options = $this->hooks->trigger('search:options', $search_type, $options, $options);
 
+		if ($this->hooks->hasHandler('search:results', $search_type)) {
```
could you change this to the regular pattern where you just fire the hook... and only if NULL do elgg_get_entities? That pattern is need so in hooks you can conditionally return the results...

hypeJunction replied:
Why wouldn't you just call ege in your hook? Just wondering

jdalsem replied:
because then i would assume what the default behaviour is... i know of multiple plugins that can be active at the same time that could take over the search results (or modify them) but determine so (conditionally) in the results hook...

jdalsem commented on docs/guides/search.rst:
```diff
@@ -29,21 +29,21 @@ In addition to all parameters accepted by ``elgg_get_entities()``, ``elgg_search
-		'fields' => [
-		   'metadata' => ['description'],
-           'annotations' => ['location'],
+    $options = [
```
i believe this should all be tabs... am i right? Maybe a misconfigured rst editor?

hypeJunction replied:
RST sucks

jdalsem commented:
almost done with this PR... just some small changes needed... and the the test am i right?

hypeJunction commented:
Yes, tests and whatever comes up during testing. I am sure there are bugs eager to be found

jdalsem commented:
> Afterwards there is only one more thing I would like to get into 3.0

is it already on the milestone?

hypeJunction commented:
> is it already on the milestone?

No. I am hoping the get a decent router in and get rid of resource views. They drive me nuts.

jdalsem commented:
> No. I am hoping the get a decent router in and get rid of resource views. They drive me nuts.

might be too big a change to do for 3.0... maybe discuss this elsewhere

hypeJunction commented:
> might be too big a change to do for 3.0... maybe discuss this elsewhere

I don't want to wait another 2 years. Going to get in some BC way of doing things.

jdalsem commented:
> Going to get in some BC way of doing things.

If it is BC it can go in a minor without a problem

hypeJunction commented:
> If it is BC it can go in a minor without a problem

Yes, but then we are stuck with the current resource views.

hypeJunction commented:
Should be done with tests layer today. On a second thought, I don't want to start with routing if there is time pressure. I have a ton work, which I should be focusing on instead.

hypeJunction committed:
feat(search): adds a new core search service

Adds a new search service that behaves similar to elgg_get_entities()

BREAKING CHANGE
Most of search plugin views have been removed or updated
Tag search has been combined with entity search.
Some of the search behavior may have changed, search hooks have been deprecated.

Fixes #7392
Fixed #11330
Fixes #11274
Fixes #7062
Fixed #3549

hypeJunction commented:
Can anyone tell me WFT Sphinx wants from me? https://travis-ci.org/Elgg/Elgg/jobs/313131505#L615

jeabakker commented:
I believe sphinx can't handle tabs in code blocks. It needs spaces .....

Op 7 dec. 2017 20:47 schreef "Ismayil Khayredinov" <notifications@github.com
>:

> Can anyone tell me WFT Sphinx wants from me? https://travis-ci.org/Elgg/
> Elgg/jobs/313131505#L615
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Elgg/Elgg/pull/11447#issuecomment-350074514>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA11JhH7Vr7wpyaGBbfB3EA6Py_p0lSuks5s-EDYgaJpZM4QyFHl>
> .
>

hypeJunction commented:
> I believe sphinx can't handle tabs in code blocks. It needs spaces .....

It references a different line... Let me try doing the opposite, replacing all tabs with 4 spaces.

hypeJunction commented:
Same shit. Still fails.

jdalsem commented:
> On a second thought, I don't want to start with routing if there is time pressure. I have a ton work, which I should be focusing on instead.

smart, i think it is another time sink

We will help you get this PR in today

jdalsem commented:
🍻
